feat(fxa-auth-server): Setup fluent localizer to localize our emails

Yogita98 opened the pull request:
## Because

- We want to replace our current localizer with fluent

## This pull request

- Includes scripts and packages that can help us setup fluent in fxa-auth-server

## Issue that this pull request solves

Closes: #9257

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [ ] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

Yogita98 commented on packages/fxa-auth-server/lib/senders/email.js:
```diff
@@ -416,7 +419,22 @@ module.exports = function (log, config) {
+      const { localizedHTML, localizedText, localizedSubject } =
+        await this.fluentLocalizer.localizeEmail(
+          message.template,
+          'hello-world'
```
@jodyheavener @LZoog Temporarily, I'm passing this string as a subject since I wasn't sure where exactly it's being created

Yogita98 commented on packages/fxa-auth-server/lib/senders/emails/locales/en-US.ftl:
```diff
@@ -0,0 +1,5 @@
+hello-world = Hello, world!
```
Let me know your thoughts on the directory structure

Yogita98 commented on packages/fxa-auth-server/lib/senders/emails/templates/cadReminder.ts:
```diff
@@ -28,10 +28,11 @@ export const render = () => {
             <mj-text css-class="secondary-text"><span>Mozilla. 2 Harrison St, #175, San Francisco, CA 94105</span></mj-text>
             <mj-text css-class="secondary-text"><span>Mozilla Privacy Policy</span></mj-text>
             <mj-text css-class="secondary-text"><span>Firefox Cloud Terms of Service</span></mj-text>
+            <mj-text><h3 data-l10n-id="hello-world">Is this localized?</h1></mj-text>
```
[fluent/dom](https://github.com/projectfluent/fluent.js/tree/master/fluent-dom) localizes the html elements that have `data-l10n-id` as an attribute. Here, it is able to localize the `h3` tag but nothing happens in case of `mj-text`. And while we can register [custom elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) by extending the HTMLElement class, all mj-elements ultimately get converted over to html.

Yogita98 commented on packages/fxa-auth-server/lib/senders/emails/templates/index.ts:
```diff
@@ -1 +1 @@
-export * as cadReminder from './cadReminder';
+export * as cadReminderMJML from './cadReminder';
```
Exporting it as `cadReminderMJML` since @vbudhram has included a regex pattern in [this PR](https://github.com/mozilla/fxa/pull/9808/commits/3c77c9039b3c6ed22638be9ff0798e170ac08340) and I thought we could use `MJML` as a pattern to enable feature flagging for the new emails

jodyheavener commented on packages/fxa-auth-server/lib/senders/email.js:
```diff
@@ -416,7 +419,22 @@ module.exports = function (log, config) {
 
-  Mailer.prototype.localize = function (message) {
+  Mailer.prototype.localize = async function (message) {
+    if (newEmailPattern.test(message.template)) {
```
I think Vijay's PR for feature flagging is already going to handle determining if an email should use the new or old template, so this shouldn't be necessary. Base on [this line](https://github.com/mozilla/fxa/pull/9808/files#diff-8f8d686a1e503181760ab061f610c01e8c357bd0da6620272cf9593d9a02e146R433) you should be able to (once that PR is merged) do something like:

```js
if (featureFlags.isMjmlEnabledForUser(message.email, message.template)) {
  // Needs to return `html` and `text` properties
  const localized = await this.fluentLocalizer.localizeEmail(
    message.template,
    'hello-world'
  );
  localizedEmailHtml = localized.html;
  localizedEmailText = localized.text;
}
```

And I _think_ this should allow you to just use the existing `cadReminderEmail` method and not need a new `cadReminderMJMLEmail`?

You could also define the subject Fluent key when calling `this.send` in the individual template mailer methods. Something like: `this.send({ mailSubject: 'hello-world' })`.

Finally, I see `localizeEmail` is importing and using `context` from `renderer.ts`, but I think you'll need to pass variables down to `this.fluentLocalizer.localizeEmail` instead. I think `message.templateValues` is what you'd pass down.

So... altogether, something like:

```js
if (featureFlags.isMjmlEnabledForUser(message.email, message.template)) {
  const { template, mailSubject, templateVariables } = message;
  const localized = await this.fluentLocalizer.localizeEmail(
    template,
    mailSubject,
    templateVariables,
  );
  localizedEmailHtml = localized.html;
  localizedEmailText = localized.text;
} else {
  // the old way
}
```

jodyheavener reviewed (commented):
Haven't dug too deeply in to this yet, I think we should wait until #9808 since it will greatly help us determine when to use the new or old localize operations.

Yogita98 commented on packages/fxa-auth-server/config/dev.json:
```diff
@@ -426,7 +426,7 @@
   "mjml": {
-    "enabledEmailAddress": "@mozilla.com$",
-    "templates":  []
+    "enabledEmailAddress": "@testuser.com$",
```
I figured that locally we're using this address to send emails so it's just pushed to test the changes

Yogita98 commented:
Things done as part of this PR:
1. Added support for optional rendering with/without a layout
2. Added separate files for `fxa` layout and `partials` like `automatedEmailNoAction`
3. Created `.ftl` files for `en` and `en-US` locales, will add more once the directory structure is finalized
4. Implemented `fluent` localization to localize `cadReminderFirst` email

Yogita98 commented on packages/fxa-auth-server/lib/senders/emails/locales/en-US.ftl:
```diff
@@ -0,0 +1,7 @@
+sync-reminder = Here's your reminder to sync devices.
+sync-description = It takes two to sync. Syncing another device with Firefox privately keeps your bookmarks, passwords and other Firefox data the same everywhere you use Firefox.
+
+random1 = Or, install on <a %(desktopLinkAttributes)s>another desktop device</a>.
```
Still need to dig a little bit more into the localization part, so leaving it here as is for now

jodyheavener commented on packages/fxa-auth-server/lib/senders/emails/layouts/fxa/index.ts:
```diff
@@ -0,0 +1,54 @@
+      <img
+             src="https://accounts-static.cdn.mozilla.net/product-icons/firefox-logo.png"
+             class="fxa-logo"
+            />
```
Some funky indentation, maybe more like this?

```
<img
  src="https://accounts-static.cdn.mozilla.net/product-icons/firefox-logo.png"
  class="fxa-logo"
/>
```

That's how we write it for React components at least

jodyheavener commented on packages/fxa-auth-server/lib/senders/emails/layouts/fxa/index.ts:
```diff
@@ -0,0 +1,54 @@
+            <mj-text css-class="secondary-text"><a class="link-blue" href="<% privacyUrl %>" class="blue-link">Mozilla Privacy Policy</a></mj-text>
+            <mj-text css-class="secondary-text"><a class="link-blue" href="https://www.mozilla.org/about/legal/terms/services/" class="blue-link">Firefox Cloud Terms of Service</a></mj-text>
+          </mj-column>
+        </mj-section>
```
Nit: Indentation needs correcting

jodyheavener commented on packages/fxa-auth-server/lib/senders/emails/partials/button/index.ts:
```diff
@@ -6,7 +6,7 @@ export const button = `
   <mj-section>
     <mj-column>
-      <mj-button css-class="primary-button" href="<%= baseURL %><%= anotherDeviceURL %>"><%= buttonText %></mj-button>
+      <mj-button css-class="primary-button" href="<%= baseUrl %><%= anotherDeviceUrl %>"><%= buttonText %></mj-button>
```
What if `anotherDeviceUrl` and other URLs incorporated `baseUrl` on the server side, so we don't need to pass down `baseUrl` as a variable everywhere?

jodyheavener commented on packages/fxa-auth-server/lib/senders/emails/partials/metadata.ts:
```diff
@@ -0,0 +1,26 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+export const metaData = `
+  <% if(oneClickLink){ %>
```
Nit: we'll want to maintain JS syntax consistency, since prettier doesn't pick up on ERB (at least not yet): `<% if (oneClickLink) { %>`

jodyheavener commented on packages/fxa-auth-server/lib/senders/emails/templates/cadReminder/index.ts:
```diff
@@ -2,39 +2,38 @@
+    <mj-text css-class="primary-text">It takes two to sync. Syncing another device with Firefox privately keeps your bookmarks, passwords and other Firefox data the same everywhere you use Firefox.</mj-text>
+  </mj-column>
+</mj-section>
+${button}
```
I just noticed this, up to you if you wanna address it in this PR or not: the button in the current template has a background of `#0a84ff`, and the new one uses `#0090ed`

jodyheavener commented on packages/fxa-auth-server/lib/senders/emails/fluentLocalizer.ts:
```diff
@@ -0,0 +1,97 @@
+
+      text: document.documentElement
+        .querySelector('body')
+        ?.textContent?.replace(/(^[ \t]*\n)|(^[ \t]*)/gm, ''),
```
What's this regex doing? Stripping HTML? At any rate text version is going to need some love -- not sure if we want to try and address this now or in a future PR:

![Screen Shot 2021-07-19 at 2 31 56 PM](https://user-images.githubusercontent.com/6392049/126202196-e10317db-866b-457e-af86-e0eeffb9177e.png)

Yogita98 replied:
Yeah I was trying to replace all the white spaces and empty lines and it worked fine until those hidden headers were added. I'll see if I can gracefully remove those as well in this PR itself or we can create a follow-up PR on this!

jodyheavener reviewed (commented):
I like how this has come along! Locally it works and renders correctly, just left a few comments so far

Yogita98 commented on packages/fxa-auth-server/lib/senders/emails/layouts/fxa/index.ts:
```diff
@@ -0,0 +1,54 @@
+          ‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌&nbsp;‌‌‌
+        </div>
+        <% } %>
+      <% if(typeof(sync) == "undefined" || sync == null){ %>
```
I tried to use the shortened condition, but it throws `undefined` error

Yogita98 commented on packages/fxa-auth-server/lib/senders/emails/templates/cadReminder/index.ts:
```diff
@@ -2,39 +2,57 @@
+    ${addDataL10nId('.automated-text div', 'automated-email-text')}
+    ${addDataL10nId('.install-device-link div', 'another-desktop-device')}
+    ${addDataL10nId('.install-desktop-device-link div', 'another-device')}
+    ${addDataL10nId('.privacy-url div', 'footer-privacy-url')}
```
I know the attribute names are looking weird, but I couldn't think about better ones. It'd be nice if we can set a convention for naming them

LZoog replied:
Hrm, I know you did this way since MJML doesn't allow custom attributes, but I think we had said before (maybe just once in our 3-way group chat) that we could potentially wrap text strings in `<span>` to apply the custom data-attribute to... I don't like the idea of extra DOM elements so I wondered if there was a better way, but doing this per template/per line to be extracted seems a little confusing/cumbersome (especially given the weirdness around MJML adding those `div`s or table cells). I don't necessarily have an alternative outside of suggesting `span`, but if we do keep this, we probably want to file a follow up ticket if even to just investigate how else we can do this.

Yogita98 commented on packages/fxa-auth-server/lib/senders/emails/fluentLocalizer.ts:
```diff
@@ -0,0 +1,97 @@
+  ) {
+    const htmlDocument = renderWithOptionalLayout(
+      templateName,
+      { ...variables, ...context },
```
The templateValues object could be passed down in a better way. Maybe this can be done as part of the cadReminder ticket

LZoog commented on packages/fxa-auth-server/lib/senders/emails/locales/ar.ftl:
```diff
@@ -0,0 +1,8 @@
+# cadReminderFirst 
```
Oooh I hope you didn't manually copy each of these over by hand (or I hope it didn't take too long). 😅

I believe these get automatically converted for us when we build - we have an l10n team that helps with translations. I see we only have the English version committed in `fxa-settings`:
![image](https://user-images.githubusercontent.com/13018240/126570334-676e65ff-f0b7-43e7-873b-588db85374f8.png)

I am not as familiar with Fluent workings as some others may be, but it seems like we only want to commit an FTL file per template/partial? @jodyheavener does that seem right to you? 🤔 I tried to find documentation on Fluent and how everything hooks up, but only found docs for GetText.

LZoog commented on packages/fxa-auth-server/package.json:
```diff
@@ -24,7 +24,7 @@
     "format": "prettier --write --config ../../_dev/.prettierrc '**'",
     "gen-keys": "ts-node ./scripts/gen_keys.js; ts-node ./scripts/oauth_gen_keys.js; ts-node ./scripts/gen_vapid_keys.js",
-    "write-emails": "node -r ts-node/register ./scripts/write-emails-to-disk.js",
+    "write-emails": "yarn emails-scss && node -r ts-node/register ./scripts/write-emails-to-disk.js",
```
Related to this comment: https://github.com/mozilla/fxa/pull/9923/files#r677542177

I thought `yarn write-emails` was for the old templates? When I try to run this, I get:

![image](https://user-images.githubusercontent.com/13018240/127179792-96e38fdf-82f7-4605-9e75-4acd06cb771e.png)

Is there another step, or are we no longer using `write-templates`?

I think we also want to take use of `yarn` out and use `npm run` instead.

On a related note, CI is failing with this (not sure if you can see):
<img width="1402" alt="image" src="https://user-images.githubusercontent.com/13018240/127180633-e9b9528c-2bff-4230-b952-a93b1f32b1e8.png">

LZoog commented on packages/fxa-auth-server/lib/senders/emails/layouts/fxa/index.scss:
```diff
@@ -0,0 +1,52 @@
+
+@use '../../global.scss';
+
+.hidden-div {
```
Maybe we can just call this `hidden`, I know it's specifically on a `div` here but it feels more versatile/like our other classes with that naming convention.

vbudhram commented on packages/fxa-auth-server/lib/senders/emails/fluentLocalizer.ts:
```diff
@@ -0,0 +1,91 @@
+    return {
+      html: document.documentElement.outerHTML,
+
+      text: document.documentElement
```
I think it might be better to do text rendering in a follow up PR. In meantime maybe pass the old text template to the email sender?

vbudhram commented on packages/fxa-auth-server/lib/senders/emails/fluentLocalizer.ts:
```diff
@@ -0,0 +1,91 @@
+      }
+    );
+
+    async function fetchResource(locale: string) {
```
I believe this will attempt to fetch and merge ftl files on every attempt to localize an email. We could store this value in a Map and have an initalize function to do the work. Similar to https://github.com/mozilla/fxa/blob/5f63d848b9bbed1a7f478b671a018a6557d13b75/packages/fxa-auth-server/lib/senders/templates/index.js#L56

Yogita98 replied:
Ohh yes, good point!

vbudhram commented on packages/fxa-auth-server/bin/mjml-server.ts:
```diff
@@ -5,9 +5,9 @@
 import stream from 'stream';
 import net from 'net';
-import { render } from '../lib/senders/emails/renderer';
+import { renderWithOptionalLayout } from '../lib/senders/emails/renderer';
```
I think you will also need to update https://github.com/mozilla/fxa/blob/8a5e52f0d05892b0bcdb1d44753e5d37bc856e46/packages/fxa-auth-server/lib/senders/emails/render-to-disk.ts#L18

vbudhram reviewed (commented):
@Yogita98 Just a few comments, but this looks really good. Storybook views are pretty neat!

<img width="838" alt="Screen Shot 2021-07-27 at 2 22 35 PM" src="https://user-images.githubusercontent.com/1295288/127207464-0b6fbf31-c6a9-4b6c-b00f-2fcc8414fcfd.png">

dannycoates commented:
I'm about to merge #9975 that'll probably make a merge conflict in `yarn.lock` here. To resolve it you should be able to rebase this branch off of `main` and run `yarn install`. Yarn _should_ fix the conflict in `yarn.lock` automatically. If `package.json` also has a conflict you should manually fix that after the rebase but before running `yarn install`. If things get too messed up I can help you fix it.

vbudhram commented on packages/fxa-auth-server/lib/senders/email.js:
```diff
@@ -434,14 +436,21 @@ module.exports = function (log, config) {
+      );
+      localizedEmailHtml = localized.html;
+      // pass old plaintext version of the email for now
+      localized = this.templates.render(
```
I think we can remove the else block below and always render original emails. Then overwrite with mjml templates as needed.

dannycoates commented:
@Yogita98 this patch should fix CircleCI

Download it and run `git apply yarnlock.patch.txt` (or just look at it and copy over the new hashes)

[yarnlock.patch.txt](https://github.com/mozilla/fxa/files/6901876/yarnlock.patch.txt)

Yogita98 commented:
@dannycoates Thank you very much for helping us with this, really appreciate you taking out time :)

vbudhram reviewed (approved):
@Yogita98 PR looks good :)

Strangely some of our content-server tests failed to acquire a lock? I've restarted them 🤞🏽

adevade16 commented:
I can see all the tests passing now, not sure what was causing the failures before 🤔

Yogita98 commented:
I can see the tests passing all of a sudden, not sure what was causing the failures before 🤔
